ui/src/main/java/com/kumuluz/ee/openapi/mp/ui/filters/SwaggerUIFilter.java

```diff
@@ -58,26 +96,150 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-
         } else {
             filterChain.doFilter(servletRequest, servletResponse);
         }
     }
 
+    private String getStaticRedirect() {
+        return uiPath + URL_QUERY + specUrl + OAUTH_QUERY + oauth2RedirectUrl;
+    }
+
+    private String createRedirectPathFromRequest(HttpServletRequest httpServletRequest) {
+        try {
+            Request request = ((Request) httpServletRequest);
+            // gather URL and other information from http request
+            String rootPath = getRootPath(request);
+            URI rootURI = URI.create(getRootUrl(request) + rootPath);
+
+            // update server list if enabled
+            if (isServerListUpdateEnabled) {
+                updateServerList(rootURI);
+            }
+
+            String rootUrl = rootURI.toString();
+            return rootPath + uiPath + URL_QUERY + rootUrl + specPath + OAUTH_QUERY + rootUrl + uiPath + OAUTH_HTML_PAGE;
+        } catch (Exception e) {
+            LOG.warning("Failed to dynamically resolve redirect from request header, " +
+                                "falling back to static configuration. Reason of failure: " + e.getMessage());
+            return getStaticRedirect();
+        }
+    }
+
+    private String getRootUrl(Request request) {
+        // note: maybe compose root url from header too when isXURICheckEnable true (X-Forwarded-Host, X-Forwarded-Proto)
+        return request.getRootURL().toString();
+    }
+
+    private String getRootPath(Request request) {
+        if (isXURICheckEnabled) {
+            // retrieve X-Original-URI if available
+            String originalURI = request.getHeader(X_ORIGINAL_URI);
+            if (originalURI != null) {
+                // remove uiPath part (which is contained in X-Org... path provided by ingres for example)
+                String rootPathFromHeader = originalURI.replaceFirst(uiPath + ".*\\z", "");
+                return removeTrailingAndAddLeadingSlash(rootPathFromHeader);
+            }
+        }
+
+        return "";
+    }
+
+    /**
+     * Updates Open API servers list.
+     * If URL is already in list or it is loopback address then no action is taken.
+     */
+    private synchronized void updateServerList(URI rootURI) {
+        OpenAPI openApiDocument = getOpenAPI();
+
+        List<Server> oldServers =
+                Optional.ofNullable(openApiDocument.getServers()).orElse(Collections.emptyList());
+
+        // update server list and in case where server was add also move it to the top
+        if (isNotInServerList(rootURI, oldServers)) {
+            // note: there is possibility that simultaneous getting of openapi .json and invoking
+            // ui endpoint will result in 2 different lists of servers
+            ArrayList<Server> newServers = new ArrayList<>();
+            newServers.add(new CustomServer(rootURI.toString()));
+            newServers.addAll(oldServers);
+            openApiDocument.setServers(newServers);
+        }
+    }
+
```

Comment:
The list is not updated when using the following configuration:
* Running the application on http://localhost:8081
* Leaving @Server annotation with http://localhost:8080

The definition is loaded, however, the "Try it out" is not working because the server address is invalid.

Reply:
fixed